src/basins/edgetracking.jl

```diff
@@ -0,0 +1,194 @@
+export edgetracking, bisect_to_edge
+
+"""
+    edgetracking(ds::DynamicalSystem, u1, u2, attractors::Dict; kwargs...)
+Runs the edge tracking algorithm[^1] for a multistable DynamicalSystem `ds`.
+
+The algorithm is initialized from two states `u1`, `u2` in state space that must belong to
+different basins of attraction. The system's `attractors` are specified as a dictionary of
+`StateSpaceSet`s.
+
+## Keyword arguments
+* `eps1 = 1e-7`: bisection distance threshold
+* `eps2 = 1e-6`: trajectory divergence distance threshold
+* `maxiter = 100`: maximum number of iterations before the algorithm stops
+* `abstol = 0.0`: convergence threshold for returned edge state (distance in state space)
+* `tmax = Inf`: maximum integration time of parallel trajectories until re-bisection 
+* `dt = 0.01`: integration time step
+* `ϵ_mapper = 0.1`: `ϵ` parameter in [`AttractorsViaProximity`](@ref)
+* `verbose = false`: if true, prints info while running
+* `output_level = 2`: controls what to return (see below)
```

Comment:
I have now written the function to return a triple `edge, track1, track2`, where `edge = (track1 + track2)/2`. Of course one could argue that one only needs `track1` and `track2` but their mean `edge` is the actual output of the algorithm that represents a track along the basin boundary, so I think it makes sense that this is the first return argument. However, the pseudo-trajectories `track1` and `track2` are also useful, as they are strictly within one basin of attraction (unlike `edge`) and can be used to derive further properties. Thus, I would not throw them away but return them as additional return arguments.

To get the edge state, one could write a helper function `edgestate(edge::StateSpaceSet)` that takes the output of `edgetracking` as input. But all that function would do is take the `N` last points of `edge`, where `N=1` if the edge state is a saddle point and some larger integer if the edge state is an extended set (limit cycle, chaotic set). So perhaps this function does too little to justify adding it?

Reply:
> However, the pseudo-trajectories `track1` and `track2` are also useful, as they are strictly within one basin of attraction (unlike `edge`) and can be used to derive further properties.

Why are they useful? What are the properties they can be used to derive?

Reply:
> To get the edge state, one could write a helper function `edgestate(edge::StateSpaceSet)` that takes the output of `edgetracking` as input.

I don't understand what this function is supposed to do. You told me its input, but I also need to know the output :D . You say the function would take the last `N` points. Where does the function get `N` from? How does the function decides when `N` should be 1?

Reply:
> > However, the pseudo-trajectories `track1` and `track2` are also useful, as they are strictly within one basin of attraction (unlike `edge`) and can be used to derive further properties.
> 
> Why are they useful? What are the properties they can be used to derive?

They are useful because they consist of segments of actual trajectories of the system, unlike their mean, `edge`. These trajectories may capture the statistical properties of the saddle more accurately than `edge`, especially in the presence of multiple timescales.

Consider a fast-slow system, e.g. fast chaotic atmosphere and slow bistable ocean. If the Lyapunov time of the atmosphere is shorter than the time between two bisection steps, then the atmospheric parts of the two parallel trajectories may evolve quite differently in between these bisection steps, and averaging them to get `edge` may distort the statistics of atmospheric variability.

In many cases one will only need `edge` but I think the code should offer the option to output `track1` and `track2` as well. We could write an extra function for returning only `edge` but I find it nicer to have only one function `edgetracking`. What do you think?

Reply:
> > To get the edge state, one could write a helper function `edgestate(edge::StateSpaceSet)` that takes the output of `edgetracking` as input.
> 
> I don't understand what this function is supposed to do. You told me its input, but I also need to know the output :D . You say the function would take the last `N` points. Where does the function get `N` from? How does the function decides when `N` should be 1?

Sorry, I wasn't clear. My suggestion was to include a helper function

`edgestate(edge::StateSpaceSet, N=1) -> edge_state`,

where `edge_state` is also of type `StateSpaceSet`. Thus the user would have to specify `N`. I don't think there is a clear automatic way to determine `N`, as the transient time until the algorithm reaches the edge state is system-dependent. All the function `edgestate` would do is cut the transient initial part based on `N` provided by the user. Should this function be added for convenience or is it unnecessary?